### PR TITLE
start debugging with options

### PR DESCRIPTION
**Pull request by wandyezj**

### CODE REVIEW

**Review by TCourtneyOwen — COMMENTED** [review]
Don't we need to update the call to startDebugging in packages/office-addin-debugging/src/commands.ts as well?

**Comment by wandyezj** [comment]
> Don't we need to update the call to startDebugging in packages/office-addin-debugging/src/commands.ts as well?

This review is currently set up to only add a new function `startDebuggingWithOptions` and not make any breaking changes.

I'm happy to completely replace the existing `startDebugging` with the new signature if that's preferable.

**Comment by TCourtneyOwen** [comment]
I think let's just update the existing method and not add a new one for the sake of simplicity

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:242]: Also, need a space after `options:` before the type.

**Review by akrantz — CHANGES_REQUESTED** [review]

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:176]: I cannot remember the differences off of the top of my head. I'm just wondering the best way to do this.

**Comment by akrantz** [comment]
Rather than use the `master` branch in your fork, you should make the a named topic branch and use that.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:279]: I think there is an extra blank line or two here that you can remove. You should just have one blank line between variable declarations and the code after.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:274]: Is there a reason to rename this from `enableLiveReload` to just `liveReload`? I'd prefer to keep the original name unless some reason to rename it.

**Comment by akrantz** [comment]
I think there might be other callers of the startDebugging() function, perhaps in test code. Can you please just double-check?

**Review by akrantz — CHANGES_REQUESTED** [review]
Looking good, but there are some minor fixes needed. Thanks.

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts]: Would it be ok to enable prettier for this repository?

I've found prettier quite convenient to maintaining code style.

**Comment by wandyezj** [comment]
> Rather than use the `master` branch in your fork, you should make the a named topic branch and use that.

I was looking at the README in this repository root, and I couldn't find the document explaining the preferred way of making contributions. 

Would you be able to link me to the appropriate document?

**Comment by akrantz** [comment]
That's the standard way in Github to work so we didn't document this specifically in the repository, but I can see how explaining better how others can contribute would be good. Let's do this in a separate PR. Basically, we you just create a branch in your fork, make the PR from the branch, and then we merge.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts]: We could but it should be done as a separate pull request. We could even merge that pull request first, and then after that is merged to master, then merge it in here.

The only difference here with the monorepo is that we need to enable prettier for each package. 

There is one prettier config that we should override with is making printWidth = 120.

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts]: From a docs point of view an interface may be preferable. 

Automatic documentation generators such as [API Extractor](https://api-extractor.com/) allow individual interface members to be documented.

For Node's @types, an interface is used to describe options objects.

Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts]: Looking at the TypeScript docs, I think using `interface` is good.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts]: Actual, for prettier we should also integrate with eslint so both lint and prettier are working.

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts:274]: Yes, the original code overwrote the input variable called `enableLiveReload` generally I try to avoid overwriting input variables because it can make it confusing to debug later.

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts]: I'm happy to enable prettier for all packages pointing at a single config in another review.

**Comment by TCourtneyOwen** [comment]
@wandyezj I don't think we document that the preferred method is to create a branch off of a fork.  I suppose we could add that to the "Contributing" section in the ReadMe

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [inline packages/office-addin-debugging/src/start.ts:190]: @akrantz How come the `app` variable is allowed to be undefined? Should `app` be a required parameter?

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:274]: Thanks for clarifying that. I agree with your rationale.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:190]: No, we will prompt the user if not specified.

Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:190]: When you use the CLI, you can do "office-addin-debugging start <manifest>" and it if there is more than one app in the manifest, it will prompt for which one, or the user can add `--app excel` to specify. If there's only one app in the manifest, it doesn't bother prompting. The VS Code configs specify the app, since you need a different launch target for each app.

Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:190]: I don't think you should try to change whether or not undefined is allowed for particular parameters here or the defaults. If you see some that you really think we should change, we can discuss it.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:231]: "If false, start without debugging; if true or undefined, start with debugging enabled."

Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:231]: Please make sure that it continues to have the behavior to enable debugging if undefined.

File: packages/office-addin-debugging/src/start.ts
Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts:190]: I see, perhaps the documentation should be updated to specify that if `app` is undefined it will prompt to specify the application?

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:176]: Consider removing the back ticks and making it `startDebugging() options` or `Options for startDebugging() function.`

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:258]: nit: I think this blank line can be removed.

**Review by akrantz — CHANGES_REQUESTED** [review]
Almost there. Some minor tweaks suggested.

**Comment by akrantz** [comment]
@wandyezj Could go through the comments and mark them resolved if you've addressed them? I think there are a few comments I made that are still outstanding.

**Comment by akrantz** [comment]
Also, you should merge master into this.

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:176]: @wandyezj Let me know if you disagree with this, but I don't see a change here. Did you want to keep the back ticks?

Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts:176]: From what I've observed in other documentation it's standard to use backticks around the names of referred to functions or objects.

Comment (akrantz) [reply packages/office-addin-debugging/src/start.ts:176]: Interesting. I haven't noticed this myself but would like to understand if this would be a better pattern to follow. Does this affect how it appears in VS Code IntelliSense?

**Review by akrantz — APPROVED** [review]

File: packages/office-addin-debugging/src/start.ts
Comment (akrantz) [inline packages/office-addin-debugging/src/start.ts:183]: Joe, based on our discussion earlier, should some of these be required parameters for the startDebugging() function instead of in the options object?

Comment (wandyezj) [reply packages/office-addin-debugging/src/start.ts:183]: In this case, since there are so many options, I think it make sense to keep it as is.

**Comment by TCourtneyOwen** [comment]
Before we merge this, I want to get a clear sense if this is a breaking change (i.e. will will need to update other template and tooling to accommodate this change. I am OK with a breaking change but just want to know what, if any, additional changes we will need to make to accommodate this change

**Comment by akrantz** [comment]
When this gets published, the major version of this package should be updated.

Other packages which consume this will need to the latest version, and then update to call StartDebugging() with the options object instead of just the parameters. `startDebugging(manifestPath, etc.)` gets replaced with:

`startDebugging({ manifestPath, enableLiveReload: false, etc. });`

It's not a difficult change.

**Pull request merged**
